Add an appendix describing Zyhybrid usage models (#1013)
riscv/riscv-cheri: src/cheri/app-hybrid-usage.adoc

[[section_hybrid_usage_models]]
== Hybrid Mode Usage Models

The `{cheri_default_ext_name}` extension enables a hart to switch between {cheri_cap_mode_name} and {cheri_int_mode_name}.

> This capability supports various usage models that facilitate the co-existence of and transition between non-CHERI software and pure-capability CHERI software.

Reviewer: The first use of the word capability in this sentence is a different meaning to the second, which can be confusing.

Author: Thanks for pointing this out, I'll rephrase it.

> === High-Level Overview of Compatibility

Reviewer 2: This section repeats what I've already put in #1006 
Which I guess is ok if they're fully in sync, but could just use some xrefs instead. They don't like repeated text.

Author: Thanks, missed that part of #1006

Author: I decided to keep it here since this makes the appendix a bit more self-contained.

There are three operating modes on a RVY-enabled core:

RVI/RVE mode (`misa.Y`=0 or `__x__envcfg.{CRE}`=0)::
  The base ISA is RVI or RVE.
  There is no access to extended YLEN registers.

>   Custom-1 through Custom-3 remain available for vendor extensions.

Reviewer 2: ```suggestion
  _custom-1_ through _custom-3_ remain available for vendor extensions.
```

> RVY pure-capability mode (`misa.Y`=1 and `__x__envcfg.{CRE}`=1, and <<pcc>>.<<m_bit>> = {CAP_MODE_VALUE})::

Reviewer 2: ```suggestion
RVY pure-capability mode (`misa.Y`=1 and `__x__envcfg.{CRE}`=1, and <<pcc>>.<<p_bit>> = {CAP_MODE_VALUE})::
```

>   The base ISA is RVY, and Custom-3 is used for RVY instructions.

Reviewer 2: custom-1 to custom-3

although we haven't used 1+2 yet, they are still in the same category - and custom instructions in 1,2 and 3 all disappear as soon as the base is RVY

>   The base ISA is RVY, and Custom-3 is used for RVY instructions.

Reviewer 2: ```suggestion
  The base ISA is RVY, and _custom-1/2/3_ are remapped to *{cheri_base_ext_name}-1/2/3*.
```

assuming #1019 is acceptable of course 😸

  All loads and stores use a YLEN base register for permissions and bounds checks.

> RVY in RVI compatibility mode (`misa.Y`=1 and `__x__envcfg.{CRE}`=1, and <<pcc>>.<<m_bit>> = {INT_MODE_VALUE})::

Reviewer 2: m_bit -> p_bit in #1006

Reviewer 2: Let's merge #1006 and then update this one

> RVY in RVI compatibility mode (`misa.Y`=1 and `__x__envcfg.{CRE}`=1, and <<pcc>>.<<m_bit>> = {INT_MODE_VALUE})::

Reviewer 2: ```suggestion
RVY in RVI compatibility mode (`misa.Y`=1 and `__x__envcfg.{CRE}`=1, and <<pcc>>.<<p_bit>> = {INT_MODE_VALUE})::
```

Reviewer 2: do we call it `compatibility mode`?

> RVY in RVI compatibility mode (`misa.Y`=1 and `__x__envcfg.{CRE}`=1, and <<pcc>>.<<m_bit>> = {INT_MODE_VALUE})::

Reviewer 2: ```suggestion
RVY in RVI compatibility mode (`misa.Y`=1 and `__x__envcfg.{CRE}`=1, and <<pcc>>.<<p_bit>> = {INT_MODE_VALUE})::
```

>   The base ISA is RVY, and Custom-3 is used for RVY instructions.

Reviewer 2: ```suggestion
  The base ISA is RVY, and *{cheri_base_ext_name}-1/2/3_ are used for RVY instructions.
```

  However, all loads and stores use a XLEN base register and check <<ddc>> for permissions and bounds.
  This mode requires `Zyhybrid` support.
  Some implementations, like CHERIoT, will not support this mode.

=== Example Usage Models

Purecap Kernel with 64-bit Plain RISC-V Programs::
  In this model, a CHERI-aware operating system (the kernel) executes in {cheri_cap_mode_name}, benefiting from full memory safety for kernel operations.
  User applications may be standard 64-bit RISC-V programs executing in {cheri_int_mode_name}.
  The kernel can decide based on the ELF header whether to start the program with CHERI enabled and in pure-capability mode or completely disabled.
+
To set up this environment and handle transitions, the following sequence is typical:
+
.  **Initialize <<ddc>> for User Space:** The kernel creates a data capability with read and write permissions.
  Its bounds are restricted to the user program's data segment.
  This capability is written to the <<ddc>> CSR.
.  **Setup <<pcc>> for User Space:** The kernel creates a capability for the user program entry point.

> It sets the <<m_bit>> to {cheri_int_mode_name} so the processor starts in integer mode.

Reviewer 2: ```suggestion
It sets the <<p_bit>> to {cheri_int_mode_name} so the processor starts in {cheri_int_mode_name}.
```

this needs replacing in a few places

This capability is written to <<sepc_y,sepc>>.
.  **Enter User Mode:** The kernel executes the `sret` instruction.
This installs the capability from <<sepc_y,sepc>> into <<pcc>>, switching the mode to {cheri_int_mode_name}.
.  **Return to Kernel Mode:** When the user program makes a system call (via `ecall`) or an error occurs, the hardware saves the current <<pcc>> into <<sepc_y,sepc>>.
It then loads the kernel's exception vector capability from <<stvec_y,stvec>> into <<pcc>>.

> Since <<stvec_y,stvec>> was set up with its <<m_bit>> set for pure-capability execution, this switches the execution mode back to {cheri_cap_mode_name}.

Reviewer 2: ```suggestion
Since <<stvec_y,stvec>> was set up with its <<p_bit>> set for pure-capability execution, this switches the execution mode back to {cheri_cap_mode_name}.
```

+
NOTE: To allow mixing plain RISC-V and pure-capability RISC-V user applications, the kernel must either context-switch the state of `senvcfg.{CRE}`, or run the plain RISC-V applications in RVI compatibility mode, in which case the custom opcode space reserved by {cheri_base_ext_name} is not available to userspace.
+
[plantuml, target="kernel-user-transition", format="svg"]
....
@startuml
box "Kernel Mode (Purecap)" #LightCyan
participant "Kernel" as K
end box

box "User Mode (Integer)" #LightYellow
participant "User App" as U
end box

K -> K: Prepare User DDC & PCC
K -> K: Write target PCC to sepc\n(mode bit = Integer)
K -> U: Execute ""sret"" (Switches to Integer Mode)
U -> U: Execute App code
U -> K: Trigger ""ecall"" or Exception (Switches to Purecap Mode)
K -> K: Handle System Call

@enduml
....

Plain RISC-V Program with a Purecap Library Sandbox::
  A large non-CHERI application (executing in {cheri_int_mode_name}) may wish to use a specific library that has been recompiled for CHERI (executing in {cheri_cap_mode_name}) to process untrusted data (e.g., an image or video codec).
+
To set up this environment and handle transitions, the following sequence is typical:
+

> .  **Prepare Library Capabilities:** A dynamic linker or runtime creates a capability for the library's entry point with the <<m_bit>> set to {cheri_cap_mode_name}.

Reviewer 2: ```suggestion
.  **Prepare Library Capabilities:** A dynamic linker or runtime creates a capability for the library's entry point with the <<p_bit>> set to {cheri_cap_mode_name}.
```

.  **Invoke the Library:** The main program (in integer mode) calls the library using <<JALR_CHERI>>.
This installs the capability into <<pcc>> and switches the mode to capability mode.
It also saves a return capability in the link register.
.  **Return from the Library:** When finished, the library returns by jumping to the return capability in the link register.
This switches the mode back to integer mode.
+
[plantuml, target="lib-sandbox-transition", format="svg"]
....
@startuml
box "Main Program (Integer)" #LightYellow
participant "Main" as M
end box

box "Library (Purecap)" #LightCyan
participant "Lib" as L
end box

M -> M: Create Library Capability\n(mode bit = Purecap)
M -> M: Set DDC to NULL
M -> L: Call via ""JALR_CHERI"" (Switches to Purecap Mode)
L -> L: Execute Sandboxed Code
L -> M: Return via return capability (Switches to Integer Mode)
M -> M: Restore DDC

@enduml
....

Pure-capability RISC-V Program Calling into a Non-CHERI Library::
  Conversely, a new pure-capability application may need to link against a non-CHERI binary-only library for which source code is unavailable.
+
To set up this environment and handle transitions, the following sequence is typical:
+

> .  **Prepare Non-CHERI Code Capability:** The pure-capability program creates a capability for the non-CHERI library's functions with the <<m_bit>> set to {cheri_int_mode_name}.

Reviewer 2: ```suggestion
.  **Prepare Non-CHERI Code Capability:** The pure-capability program creates a capability for the non-CHERI library's functions with the <<p_bit>> set to {cheri_int_mode_name}.
```

participant "C++ Runtime" as R
end box

note over J: Running in Integer Mode

== Accessing C Runtime Data ==
J -> J: Execute ""MODESW_CAP"" (Switches to Capability Mode)
J -> J: Access runtime data using\nexplicit capabilities
J -> J: Execute ""MODESW_INT"" (Switches to Integer Mode)

== Calling Runtime Helpers ==
J -> R: Call helper (Sentry Capability) (Switches to Capability Mode)
R -> R: Execute safe runtime code
R -> J: Return to JIT code (Switches to Integer Mode)

@enduml
....

=== CSRs and Registers Involved in Mode Transitions

When transitioning between modes, certain capability registers and CSRs must be configured to ensure correct execution and authorization:

<<ddc>> (Default Data Capability)::
This CSR must be written with a capability covering the data memory that the code in {cheri_int_mode_name} is authorized to access.
Memory accesses in {cheri_int_mode_name} are implicitly authorized by <<ddc>>.

Exception Return Capability Registers (`sepc` or `mepc`)::
When a privileged environment prepares to transition to a lower privilege mode via an exception return instruction, it must write the target code capability to `sepc` or `mepc`.

> This capability must have its <<m_bit>> set to the desired execution mode ({cheri_cap_mode_name} or {cheri_int_mode_name}).

Reviewer 2: ```suggestion
This capability must have its <<p_bit>> set to the desired execution mode ({cheri_cap_mode_name} or {cheri_int_mode_name}).
```

The exception return instruction installs this capability into <<pcc>>, thereby setting the mode.

Trap Vector Registers (<<stvec_y,stvec>> or <<mtvec_y,mtvec>>)::
These registers hold the capability for the exception vector.
  When an exception or interrupt occurs, the hardware installs this capability into <<pcc>>, thereby setting the mode for the trap handler.

>   For example, if <<stvec_y,stvec>> was set up with its <<m_bit>> set to {CAP_MODE_VALUE}, the processor will switch to {cheri_cap_mode_name} on entry to the trap handler.

Reviewer 2: ```suggestion
  For example, if <<stvec_y,stvec>> was set up with its <<p_bit>> set to {CAP_MODE_VALUE}, the processor will switch to {cheri_cap_mode_name} on entry to the trap handler.
```

=== CHERI Enable Bits in Setup (when changing privilege levels)

In addition to setting up capabilities and mode bits, the execution environment must manage the CHERI enable bits provided by the privileged architecture:

Per-Privilege Enable Bits::
The privileged architecture defines enable bits (often denoted as `{CRE}`) in CSRs such as `misa`, `menvcfg`, and `senvcfg` to control CHERI availability in each privilege mode (see <<section_cheri_disable>>).

Reset State::
On reset, CHERI is disabled for all modes to ensure 100% backwards compatibility with standard RISC-V software.
This behavior applies to cores that support execution in both {cheri_cap_mode_name} and {cheri_int_mode_name}.

> Implementations that do not support RVI/RVE base ISAs and always run with a RVY base ISA (such as CHERIoT) cannot execute standard RVI code and therefore start up with CHERI enabled.

Reviewer 2: ```suggestion
```

I don't think we need this sentence, this whole section is about Zyhybrid support and it says elsewhere that CHERIoT doesn't support that.

Enabling CHERI::
A higher privilege mode must set the appropriate enable bits to allow a lower privilege mode to use CHERI instructions or switch between {cheri_cap_mode_name} and {cheri_int_mode_name}.

Disabled CHERI with Bounds Enforcement::
If CHERI is disabled for a specific privilege mode, that mode cannot execute CHERI instructions or change modes.
However, the hardware continues to enforce the bounds defined by the last installed <<pcc>> and <<ddc>>.
This allows a kernel to run non-CHERI-aware code in a strict integer environment while still confining it within capability bounds.

=== Transition Instructions and Frequency

Transitions between modes are performed using specific instructions:

<<MODESW_INT>> and <<MODESW_CAP>>::
These instructions are used to switch between modes without changing the control flow.
They are typically used when entering or leaving a section of code that requires a different mode.
<<JALR_CHERI>>::

> When jumping to a target capability with the <<m_bit>> set differently from the current mode, the jump performs a mode switch.

Reviewer 2: ```suggestion
When jumping to a target capability with the <<p_bit>> set differently from the current mode, the jump performs a mode switch.
```
